src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs

```diff
@@ -27,6 +27,7 @@ public Utf8JsonReader(in ReadOnlySequence<byte> jsonData, bool isFinalBlock, Jso
             _buffer = jsonData.First.Span;
 
             _isFinalBlock = isFinalBlock;
+            _isInputSequence = true;
 
             // Note: We do not retain _bytesConsumed or _sequencePosition as they reset with the new input data
```

Comment:
Why do we store _sequencePosition if it doesn't affect subsequent uses?  Is it just about reporting the information back to the consumer?

Reply:
Yes.

Reply:
Is that an indication it shouldn't be on the state object? Sounds like it's making the state type bigger (which will in turn make objects like state machines bigger), but the main purpose of the type is to support flowing state from one reader to another. If it's not needed for that purpose, shouldn't it instead just be a property on the reader?

Reply:
The same can be said about `BytesConsumed`.

However, in practice, `BytesConsumed` (for the async stream/span case) and `Position` (for the async PipeReader case) usually need to survive async/await boundaries so that the caller can do appropriate re-slicing/etc. If they are accessible on the reader (only), then the caller has to explicitly grab them from the reader before the next call. Keeping on the `JsonReaderState` makes it easier for the caller.

```C#
ReadResult result = await _pipe.Reader.ReadAsync();
ReadOnlySequence<byte> data = result.Buffer;
state = ProcessData(data, isFinalBlock, state); // contains the ref struct Utf8JsonReader
_pipe.Reader.AdvanceTo(state.Position);
```

Reply:
> The same can be said about BytesConsumed.

Yes.

> usually need to survive async/await boundaries so that the caller can do appropriate re-slicing/advancing/etc

It doesn't survive an await boundary in your example.  When would it need to?

> Keeping them on the JsonReaderState makes it easier for the caller.

Is this type about usability or about performance?

If the former, man do we have a bunch of things to fix ;)

If the latter, then why are we ok leaving perf on the table?  Looks like this would shave 24 bytes off of a JsonReaderState, which means 24 bytes off of every state machine it's stored in that ends up on the heap.

Don't hold up this PR for it.  But we should really examine what's necessary to live on JsonReaderState and what doesn't have to.

Reply:
> It doesn't survive an await boundary in your example. When would it need to?

You are right. I meant the method responsible for reading more data needs it for subsequent reads and it doesn't have access to the `Utf8JsonReader`.

> Looks like this would shave 24 bytes off of a JsonReaderState, which means 24 bytes off of every state machine it's stored in that ends up on the heap.

That's a good point. Filed an issue: https://github.com/dotnet/corefx/issues/35252

Reply:
> Is this type about usability or about performance?

Is the cost of returning a value tuple usually negligible (especially when compared to storing a larger state machine on the heap)?